facebookresearch/ParlAI
# Decoder-Only Transformer

Supporting decoder-only transformer training:
```
parlai train_model --model transformer/decoder
```

**Summary of changes**
- New `TransformerDecoderLayer` without encoder attention
- New `TransformerDecoder` that concatenates input and "encoder state"
        - Needed to pass the dictionary to `TransformerDecoder` to properly do concatenation
	- Right-pad the label, left-pad the context
- New `PassThroughEncoder`, similar to `IdentityLayer` but compatible with TransformerEncoder API
	- There are many assumptions about there being an "encoder" in `TorchGeneratorAgent`, so the path of most code reuse led to using a dummy encoder to satisfy those assumptions
- New `DecoderAgent` to override `build_model`
- Extract label from scores to calculate loss (done at end of `TransformerDecoderOnly.forward`)
- Only use last `query_len` rows from incremental attention to account for incremental step from context to first generated token
- Added `DecoderIncrState` and `DecoderLayerIncrState` type aliases
- Added unit tests

**Summary of structural changes in `decoder.py`**
![decoder_only_refac](https://user-images.githubusercontent.com/1254056/151404227-b388f4e9-fc5c-419d-815c-bf77ec9563f7.png)



**Testing steps**
```
python -m pytest tests/test_transformers.py
```
<img width="1790" alt="Screen Shot 2022-02-08 at 12 37 03" src="https://user-images.githubusercontent.com/1254056/153062569-649001a0-2502-46d2-85af-f51c0a6400fe.png">

```
python -m pytest tests/test_light_whoami.py
```
<img width="1791" alt="Screen Shot 2022-02-08 at 12 12 04" src="https://user-images.githubusercontent.com/1254056/153062686-e70efedd-1505-4a45-8d46-dd6900b3f635.png">

Also ran some small training runs locally to sanity check:
```
TF_ARGS="--embedding-size 128 --ffn-size 512 --batchsize 16 --eval-batchsize 16 --model-parallel true --variant prelayernorm --n-heads 8 --n-positions 512 --activation gelu --text-truncate 256 --label-truncate 128 -lr 7e-05 --lr-scheduler invsqrt --optimizer adam --warmup_updates 1000 -vp 10 -vmt ppl -vmm min --load-from-checkpoint false -vstep 500 --validation-max-exs 1000 -tstep 150000 --log-every-n-secs 30 --update-freq 1 --dynamic-batching full"
parlai train_model -t convai2 --model transformer/decoder --model-file /tmp/test_decoder_only_model --n-layers 2 $TF_ARGS
parlai train_model -t convai2 --model transformer/generator --model-file /tmp/test_enc_decoder_model --n-encoder-layers 1 --n-decoder-layers 1 $TF_ARGS
```

**Benchmark Comparison with Encoder-Decoder**
Both trained to 10k steps with 120M parameters (8 layers) on 2x Quadro GP100 GPUs. The encoder/decoder is closer to 140M parameters due to the cross-attention.
```
TF_ARGS="--embedding-size 1024 --ffn-size 4096 --batchsize 16 --eval-batchsize 16 --model-parallel true --variant prelayernorm --n-heads 16 --n-positions 512 --activation gelu --text-truncate 256 --label-truncate 128 -lr 7e-05 --lr-scheduler invsqrt --optimizer adam --warmup_updates 1000 -vp 10 -vmt ppl -vmm min --load-from-checkpoint false -vstep 1000 --validation-max-exs 1000 -tstep 10000 --log-every-n-secs 30 --update-freq 1 --dynamic-batching full --fp16 true"
```
```
parlai train_model -t internal:new_reddit --model transformer/decoder --model-file /tmp/test_decoder_only_model --n-layers 8 $TF_ARGS
...
13:17:37 | time:6960s total_exs:433912 total_steps:9950 epochs:0.00 time_left:35s
    clen  clip  ctpb  ctps  ctrunc  ctrunclen  exps  exs  fp16_loss_scalar  gnorm  gpu_mem  llen  loss        lr  ltpb  ltps  ltrunc  ltrunclen   ppl  token_acc  token_em  total_train_updates  tpb  tps   ups
   58.72     1  2603  4957  .01423      .8327 85.63 2248            131072  1.711    .2126 26.09 4.036 2.219e-05  1173  2234       0          0 56.59      .2843         0                 9950 3776 7191 1.905

13:18:04 | Stopping from Maximum LR steps
```
```
parlai train_model -t internal:new_reddit --model transformer/generator --model-file /tmp/test_enc_decoder_model --n-encoder-layers 4 --n-decoder-layers 4 $TF_ARGS
...
10:52:31 | time:5367s total_exs:431792 total_steps:9950 epochs:0.00 time_left:27s
    clen  clip  ctpb  ctps  ctrunc  ctrunclen  exps  exs  fp16_loss_scalar  gnorm  gpu_mem  llen  loss        lr  ltpb  ltps  ltrunc  ltrunclen   ppl  token_acc  token_em  total_train_updates  tpb   tps  ups
   64.57     1  2711  6939  .02041      1.706 110.4 2156            131072  1.633    .1838 28.03 4.066 2.219e-05  1208  3094       0          0 58.33      .2803         0                 9950 3919 10033 2.56

10:52:51 | Stopping from Maximum LR steps
```
